File: ios/SchoolHelperIOS/Features/Setup/SetupViewModel.swift
import Foundation
import Combine

@MainActor
final class SetupViewModel: ObservableObject {
    @Published var draftProfile: StudentProfile
    @Published var searchQuery: String
    @Published var searchResults: [SchoolInfo] = []
    @Published var selectedSchool: SchoolInfo?
    @Published var message: String = ""
    @Published var isSearching: Bool = false

    private let repository: SchoolRepository
    private var latestSearchRequestID = 0

    init(
        initialProfile: StudentProfile,
        repository: SchoolRepository = DefaultSchoolRepository()
    ) {
        self.draftProfile = initialProfile
        self.searchQuery = initialProfile.schoolName
        self.selectedSchool = initialProfile.schoolInfo
        self.message = Self.initialSchoolMessage(for: initialProfile)
        self.repository = repository
    }

    func updateSchoolQuery(_ query: String) {
        latestSearchRequestID += 1
        searchQuery = query
        searchResults = []

        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        if let selectedSchool, selectedSchool.schoolName != trimmed {
            self.selectedSchool = nil
            message = trimmed.isEmpty ? "" : "검색 결과에서 학교를 다시 선택해 주세요."
        } else if trimmed.isEmpty {
            message = ""
        }
    }

    func searchSchoolsAfterDebounce() async {
        let trimmed = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 2 else { return }
        guard selectedSchool?.schoolName != trimmed else { return }

        do {
            try await Task.sleep(nanoseconds: 450_000_000)
        } catch {
            return
        }
        guard !Task.isCancelled else { return }
        await searchSchools()
    }

    func searchSchools() async {
        let trimmed = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 2 else {
            message = "학교 이름은 두 글자 이상 입력해 주세요."
            searchResults = []
            return
        }

        latestSearchRequestID += 1
        let requestID = latestSearchRequestID
        isSearching = true

        do {
            let schools = try await repository.searchSchools(query: trimmed)
            guard isLatestSearch(requestID: requestID, query: trimmed) else {
                return
            }
            isSearching = false
            searchResults = schools
            if schools.count == 1, let school = schools.first {
                selectSchool(school)
                message = "학교 1개를 찾았어요."
            } else if schools.isEmpty {
                message = "검색 결과가 없어요."
            } else {
                message = "검색 결과에서 학교를 선택해 주세요."
            }
        } catch {
            guard isLatestSearch(requestID: requestID, query: trimmed) else {
                return
            }
            isSearching = false
            message = error.localizedDescription
        }
    }

    func selectSchool(_ school: SchoolInfo) {
        selectedSchool = school
        draftProfile.schoolName = school.schoolName
        draftProfile.officeCode = school.officeCode
        draftProfile.schoolCode = school.schoolCode
        draftProfile.schoolKind = school.schoolKind
        searchQuery = school.schoolName
        message = "\(school.schoolName)을 선택했어요."
    }

    func updateGrade(_ grade: String) {
        draftProfile.grade = grade
    }

    func updateClassroom(_ classroom: String) {
        draftProfile.classroom = classroom
    }

    func buildProfileForSave() -> StudentProfile? {
        let trimmedQuery = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let school = selectedSchool, school.schoolName == trimmedQuery else {
            message = "학교를 검색 후 다시 선택해 주세요."
            return nil
        }
        let trimmedGrade = draftProfile.grade.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedClassroom = draftProfile.classroom.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedGrade.isEmpty, !trimmedClassroom.isEmpty else {
            message = "학년/반을 입력해 주세요."
            return nil
        }

        var profile = draftProfile
        profile.grade = trimmedGrade
        profile.classroom = trimmedClassroom
        profile.schoolName = school.schoolName
        profile.officeCode = school.officeCode
        profile.schoolCode = school.schoolCode
        profile.schoolKind = school.schoolKind
        return profile
    }

    private func isLatestSearch(requestID: Int, query: String) -> Bool {
        requestID == latestSearchRequestID &&
            searchQuery.trimmingCharacters(in: .whitespacesAndNewlines) == query
    }

    private static func initialSchoolMessage(for profile: StudentProfile) -> String {
        let schoolName = profile.schoolName.trimmingCharacters(in: .whitespacesAndNewlines)
        if !schoolName.isEmpty && !profile.hasSchoolSelection {
            return "기존 설정에 학교 코드가 없어 학교를 다시 검색해 선택해 주세요."
        }
        return ""
    }
}

File: android/app/src/main/java/com/lbs/schoolhelper/ui/setup/SetupViewModel.kt
package com.lbs.schoolhelper.ui.setup

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.lbs.schoolhelper.R
import com.lbs.schoolhelper.data.model.SchoolInfo
import com.lbs.schoolhelper.data.repository.PreferencesRepository
import com.lbs.schoolhelper.data.repository.SchoolRepository
import com.lbs.schoolhelper.data.repository.StudentInfo
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class SetupViewModel @Inject constructor(
    application: Application,
    private val preferencesRepository: PreferencesRepository,
    private val schoolRepository: SchoolRepository
) : AndroidViewModel(application) {
    private val appContext = application.applicationContext
    private val initialStudentInfo = preferencesRepository.getStudentInfo()
    private var schoolSearchJob: Job? = null
    private var latestSearchRequestId: Long = 0L

    private val _uiState = MutableStateFlow(
        SetupUiState(
            schoolQuery = initialStudentInfo.schoolName,
            selectedSchool = initialStudentInfo.takeIf { it.hasSchoolSelection() }?.toSchoolInfo(),
            searchMessage = if (initialStudentInfo.schoolName.isNotBlank() && !initialStudentInfo.hasSchoolSelection()) {
                appContext.getString(R.string.school_search_reselect_required)
            } else {
                ""
            },
            grade = initialStudentInfo.grade,
            classroom = initialStudentInfo.classroom
        )
    )
    val uiState = _uiState.asStateFlow()

    private val _messageEvent = MutableSharedFlow<Int>()
    val messageEvent = _messageEvent.asSharedFlow()

    private val _navigationEvent = MutableSharedFlow<Unit>()
    val navigationEvent = _navigationEvent.asSharedFlow()

    fun updateSchoolQuery(query: String) {
        val trimmedQuery = query.trim()
        _uiState.update { state ->
            state.copy(
                schoolQuery = query,
                selectedSchool = state.selectedSchool,
                schoolResults = emptyList(),
                searchMessage = when {
                    state.selectedSchool != null && state.selectedSchool.schoolName != trimmedQuery ->
                        appContext.getString(R.string.school_search_reselect_current_selection)
                    else -> state.searchMessage
                }
            )
        }
    }

    fun searchSchools() {
        val query = _uiState.value.schoolQuery.trim()
        if (query.length < MIN_SCHOOL_QUERY_LENGTH) {
            _uiState.update {
                it.copy(
                    isSearching = false,
                    searchMessage = appContext.getString(R.string.school_search_min_query),
                    schoolResults = emptyList(),
                    selectedSchool = it.selectedSchool
                )
            }
            return
        }

        val requestId = ++latestSearchRequestId
        schoolSearchJob?.cancel()
        schoolSearchJob = viewModelScope.launch {
            _uiState.update { it.copy(isSearching = true, searchMessage = "", schoolResults = emptyList()) }
            val result = schoolRepository.searchSchools(query)
            if (requestId != latestSearchRequestId) return@launch

            result.onSuccess { schools ->
                when {
                    schools.isEmpty() -> {
                        _uiState.update {
                            it.copy(
                                isSearching = false,
                                schoolResults = emptyList(),
                                selectedSchool = it.selectedSchool,
                                searchMessage = appContext.getString(R.string.school_search_empty)
                            )
                        }
                    }

                    schools.size == 1 -> {
                        val selectedSchool = schools.first()
                        _uiState.update {
                            it.copy(
                                schoolQuery = selectedSchool.schoolName,
                                selectedSchool = selectedSchool,
                                schoolResults = schools,
                                searchMessage = appContext.getString(R.string.school_search_single_result),
                                isSearching = false
                            )
                        }
                    }

                    else -> {
                        _uiState.update {
                            it.copy(
                                schoolResults = schools,
                                selectedSchool = null,
                                searchMessage = appContext.getString(R.string.school_search_select_result),
                                isSearching = false
                            )
                        }
                    }
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(
                        isSearching = false,
                        schoolResults = emptyList(),
                        selectedSchool = it.selectedSchool,
                        searchMessage = error.message ?: appContext.getString(R.string.school_search_error)
                    )
                }
            }
        }
    }

    fun selectSchool(school: SchoolInfo) {
        _uiState.update {
            it.copy(
                schoolQuery = school.schoolName,
                selectedSchool = school,
                searchMessage = appContext.getString(R.string.school_search_selected, school.schoolName)
            )
        }
    }

    fun updateGrade(grade: String) {
        _uiState.update { it.copy(grade = grade) }
    }

    fun updateClassroom(classroom: String) {
        _uiState.update { it.copy(classroom = classroom) }
    }

    suspend fun saveStudentInfo() {
        val state = _uiState.value
        if (state.selectedSchool == null || state.selectedSchool.schoolName != state.schoolQuery.trim()) {
            _messageEvent.emit(R.string.setup_error_school_required)
            return
        }
        if (state.grade.isBlank() || state.classroom.isBlank()) {
            _messageEvent.emit(R.string.setup_error_empty)
            return
        }

        preferencesRepository.saveStudentInfo(
            StudentInfo(
                grade = state.grade,
                classroom = state.classroom,
                schoolName = state.selectedSchool.schoolName,
                officeCode = state.selectedSchool.officeCode,
                schoolCode = state.selectedSchool.schoolCode,
                schoolKind = state.selectedSchool.schoolKind
            )
        )
        _navigationEvent.emit(Unit)
    }

    companion object {
        private const val MIN_SCHOOL_QUERY_LENGTH = 2
    }
}
